engine/sources/containerimagedns/state.go

```diff
@@ -0,0 +1,105 @@
+package containerimagedns
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/distribution/reference"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/client/llb/sourceresolver"
+	"github.com/moby/buildkit/solver/pb"
+	"github.com/moby/buildkit/util/apicaps"
+)
+
+func Image(ref string, namespace string, opts ...llb.ImageOption) llb.State {
+	r, err := reference.ParseNormalizedNamed(ref)
+	if err == nil {
+		r = reference.TagNameOnly(r)
+		ref = r.String()
+	}
+	var info ImageInfo
+	for _, opt := range opts {
+		opt.SetImageOption(&info.ImageInfo)
+	}
+
+	addCap(&info.Constraints, pb.CapSourceImage)
+
+	attrs := map[string]string{}
+	if info.resolveMode != 0 {
```

Comment:
info.resolveMode is private, but making a package-local ImageOption then goes and breaks the SetImageOption interface... do I keep pulling the thread, making copies of all these structs?

Reply:
pulled on the "reimplement" thread and it gets pretty crazy, we'd end up with a whole shadow copy of llb constraints. not exactly feasible.